### ISSUE TITLE
Calling "clear()" before "sendKeys()" causes some chars to be eaten

### ISSUE THREAD

**Issue by romaricdrigon** [facebook/php-webdriver#704]
Hello,

I could dig some more the issue from yesterday, and reproduce it.

### What are you trying to achieve? (Expected behavior)

I'm trying to log in https://the-internet.herokuapp.com/login
I'm using the following code, which reproduces the way Symfony Panther calls webdriver:
```php
$email = $driver->findElement(WebDriverBy::name('username'));
$email->clear();
$email->sendKeys('tomsmith');

$password = $driver->findElement(WebDriverBy::name('password'));
$password->clear();
$password->sendKeys('SuperSecretPassword!');

$driver->findElement(WebDriverBy::cssSelector('button[type="submit"]'))->click();
```

### What do you get instead? (Actual behavior)

Last character from `username` input is never typed, causing login to fail.

### How could the issue be reproduced? (Steps to reproduce)

The issue disappear if `$email->clear();` and `$password->clear();` are commented.

It is not a timing issue, adding a `sleep(5)` or so does not help.
The crash looks to be caused by [lines 76-93 of RemoteWebElement::clear()](https://github.com/facebook/php-webdriver/blob/community/lib/Remote/RemoteWebElement.php#L76). Commenting those or replacing "WebDriverKeys::BACKSPACE" by "WebDriverKeys::DELETE" fixed it.

Here's a minimal gist to reproduce the issue: https://gist.github.com/romaricdrigon/1c2c3d3ab99644f61ba21f8334d6fc18

### Details

* Php-webdriver version: latest `community` (`1.8.x-dev`), it does not happen in `1.7.1`
* PHP version: 7.4.0
* Selenium server version: using ChromeDriver directly, 78.0.3904.70 or 78.0.3904.105
* Operating system: reproduced both on Debian (dockerized) and macOS
* Browser used + version: Chromium 78.0.3904.108 or Chrome 78.0.3904.108

**Comment by romaricdrigon** [comment]
Note: I did a blind guess based on another issue, and tried to add `$driver->switchTo()->activeElement()`, but it did not help.

```php
$email = $driver->findElement(WebDriverBy::name('username'));
$driver->switchTo()->activeElement();
$email->clear(); // Comment to fix
$driver->switchTo()->activeElement();
$email->sendKeys('tomsmith');
```

**Comment by OndraM** [comment]
@romaricdrigon Nice job, thanks! :100: We will have a look. As you noted, the "clear" operation is emulated via actions in W3C protocol mode, so it not yet field tested. 

If this is blocking you, you may either downgrade to 1.7.1 or temporarily [disable W3C protocol](https://github.com/facebook/php-webdriver/pull/674#issue-339395010) and fall back to JsonWire.

**Comment by romaricdrigon** [comment]
Thanks, yes right now I'm using a "1.7.2" I did on a fork.

**Comment by andrewnicols** [comment]
I'm seeing this too and have been investigating it.

As I mentioned in #691,  the `clear` command causes a blur and loss of focus, so our current action of immediately sending the C+A Backspace keys does not actually work. See https://github.com/facebook/php-webdriver/pull/691#issuecomment-559883408 for some more info.

It's on my agenda to look into this more today.

**Comment by andrewnicols** [comment]
I've also found that the Ctrl+A is not actually selecting all, and it's not applying to the current field in some situations.

So what's actually happening in your example @romaricdrigon  is:
1. I clear the `username` field
1. I set the value of the username field to 'tomsmith'
1. I clear the `password` field but the Ctrl+A backspace is actually sent to the _previously_ selected element (username) and only the backspace is pressed
1. I set the value of the password

As mentioned in https://github.com/w3c/webdriver/issues/445, the clear causes the blur, and then we type, but the next thing to be sent is an `action` with Ctrl+A backspace, but there is no active focus first so the action seems to be sent  to the previously focused thing.

Frustratingly I can reproduce this in my Mink driver, but not with the same steps in the clear unit tests.

**Comment by romaricdrigon** [comment]
Thanks for investigating, and I realize it may explain some difference I saw in my tests: on macOS, the last char from previous field is removed, on Linux, the whole previous field is cleared.
So before sending Ctrl+A, we should reselect element?
Still, I don't understand why `$driver->switchTo()->activeElement();` does not work?

**Comment by andrewnicols** [comment]
On Mac the control and meta keys are swapped. Control+a on Mac is equivalent to the home key but it doesn’t seem to have that effect in webdriver.

If you take a look at the clear function you’ll also see that the key up for the control happens before the key up for the a key too. Which doesn’t make sense either.

I suspect that active element doesn’t work because it is no longer active.

On my mink driver I’ve changed it to get the length of the current value and hit backspace that many times. I did try command a, backspace, and a few other options but none work reliably.

**Comment by OndraM** [comment]
>  Still, I don't understand why $driver->switchTo()->activeElement(); does not work?

Because by just locating element you won't make it active (and the body will still be the active element). You must at first interact with the element to make it active, by eg. clicking into it..

**Comment by OndraM** [comment]
@romaricdrigon Could you please have a look in #707, whether it fixes it for you? At least the scenario from this issue description started to work for me properly.

**Comment by romaricdrigon** [comment]
#707 solves it, so I think we can close this issue once it is merged
